src/Database/LSMTree/Internal/RawOverflowPage.hs

```diff
@@ -0,0 +1,161 @@
+import           Database.LSMTree.Internal.Serialise.RawBytes (RawBytes (..))
+import qualified Database.LSMTree.Internal.Serialise.RawBytes as RB
+
+-------------------------------------------------------------------------------
+-- RawOverflowPage type
+-------------------------------------------------------------------------------
+
+-- | When a key\/op pair is too large to fit in a single disk page, the
+-- representation is split into a normal page, and one or more overflow pages.
+-- The normal 'RawPage' follows the run page format, and contains the key,
+-- optional blob reference and a prefix of the value, while the overflow pages
+-- contain the suffix of a large value that didn't fit in the normal page.
+--
+-- Each overflow page is the same size as normal pages (currently 4096 only).
+--
+data RawOverflowPage = RawOverflowPage
+                         !Int        -- ^ offset in Word8s.
+                         !ByteArray
+  deriving (Show)
+
+-- | This invariant is the same as for 'RawPage', but there is no alignment
+-- constraint. This is for two reasons: 1. we don't need alignment, because
+-- the page has no structure that we need to read with aligned memory
+-- operations; 2. we don't want alignment because we want to convert the suffix
+-- of serialised values (as 'RawBytes') into a 'RawOverflowPage' without
+-- copying, and a suffix can start at an arbitrary offset.
+--
+invariant :: RawOverflowPage -> Bool
+invariant (RawOverflowPage off ba) = and
+    [ off >= 0                            -- offset is positive
+                                          -- no alignment constraint
+    , (sizeofByteArray ba - off) >= 4096  -- there are always 4096 bytes
+    , isByteArrayPinned ba                -- bytearray must be pinned
+    ]
+
+instance NFData RawOverflowPage where
+  rnf (RawOverflowPage _ _) = ()
+
+-- | This instance assumes pages are 4096 bytes in size
+instance Eq RawOverflowPage where
+    RawOverflowPage off1 ba1 == RawOverflowPage off2 ba2 =
+      P.Vector off1 4096 ba1 == (P.Vector off2 4096 ba2 :: P.Vector Word8)
+
+-- | Create a 'RawOverflowPage'.
+--
+-- The length must be 4096 or less.
+--
+-- This function will copy data if the byte array is not pinned, or the length
+-- is strictly less than 4096.
+--
+makeRawOverflowPage
+    :: ByteArray  -- ^ bytearray
+    -> Int        -- ^ offset in bytes into the bytearray
+    -> Int        -- ^ length in bytes, must be @>= 0 && <= 4096@
+    -> RawOverflowPage
+makeRawOverflowPage ba off len
+    | len == 4096
+    , let page = RawOverflowPage off ba
+    , invariant page
+    = page
+
+    | otherwise
+    = makeRawOverflowPageCopy ba off len
+
+makeRawOverflowPageCopy
+    :: ByteArray  -- ^ bytearray
+    -> Int        -- ^ offset in bytes into the bytearray
+    -> Int        -- ^ length in bytes
+    -> RawOverflowPage
+makeRawOverflowPageCopy ba off len =
+    assert (len >= 0 && len < 4096 && len <= sizeofByteArray ba - off) $
```

Comment:
Would this function not also be used on inputs where `len == 4096` (failing this invariant), but the byte array is not pinned?

```suggestion
    assert (len >= 0 && len <= 4096 && len <= sizeofByteArray ba - off) $
```

Reply:
Yes I agree. Done.